### PR TITLE
RPC: Also serve txo from gettxout (not just utxo and mempool)

### PR DESCRIPTION
**Pull request by jtimon**
Expand gettxout to also serve spent txo if pruning or lack of -txindex don't get in the way.

If the output is spent, it will be slower to load, but there's no reason not to serve this if the data is there.

TODO (functional tests):

- [x] Test success (ideally more than one example)
- [x] Test error no txid
- [x] Test error no n for this txid
- [x] Test spent/unspent
- [x] Test txindex error
- [ ] Test prunning error

### CODE REVIEW

**Comment by sipa** [comment]
I don't think we should be expanding the impact of `-txindex`. This functionality is already available through `getrawtransaction` anyway.

**Comment by jtimon** [comment]
> I don't think we should be expanding the impact of -txindex.

Is there a problem with that? This is just RPC anyway.
I also think making -txindex more useful may increase the incentives to run an archive full node.

My reasoning (with some errors corrected below it) is here: https://github.com/bitcoin/bitcoin/pull/9806#issuecomment-313578569

> This functionality is already available through getrawtransaction anyway.

By that logic, shouldn't we just fully remove gettxout?
That would be bad for gettxoutbyscript and gettxoutbyaddress though, I think.

gettxout is faster for utxo, so if one wants a txo (not knowing if stxo/utxo a priori), one could call gettxout first and only if it fails call getrawtransation. But I think a single call would be more convenient.
If not, I guess we need getoutpointfromscript and getoupointfrom address call too to serve the full txo.
One could, for example, call gettxoutbyaddress and if it fails, also call gettxoutbyaddress and getrawtransaction. It would be more cumbersome, but it would also work.

File: src/validation.cpp [hunk 1]
```diff
@@ -955,6 +948,23 @@ bool GetTransaction(const uint256 &hash, CTransactionRef &txOut, const Consensus
...
+/**
+ * Return transaction in txOut, and if it was found inside a block,
+ * its hash is placed in hashBlock. Look in the mempool first.
+*/
+bool GetTransaction(const uint256 &hash, CTransactionRef &txOut, const Consensus::Params& consensusParams, uint256 &hashBlock, bool fAllowSlow)
+{
+    LOCK(cs_main);
+
+    CTransactionRef ptx = mempool.get(hash);
+    if (ptx)
```
Comment (promag) [inline src/validation.cpp]:
```cpp
if (ptx) {
```

File: test/functional/wallet.py [hunk 2]
```diff
@@ -19,6 +19,7 @@ def __init__(self):
         self.setup_clean_chain = True
         self.num_nodes = 4
         self.extra_args = [['-usehd={:d}'.format(i%2==0)] for i in range(4)]
+        self.extra_args[0].append('-txindex')
```
Comment (promag) [inline test/functional/wallet.py]: @jnewbery this is why I prefer isolated tests. This is unrelated to all these tests except the new.

File: src/rpc/blockchain.cpp [hunk 3]
```diff
@@ -987,6 +1016,12 @@ UniValue gettxout(const JSONRPCRequest& request)
     if (request.params.size() > 2)
         fMempool = request.params[2].get_bool();
 
+    bool include_spent = false;
+    if (request.params.size() > 3) {
+        include_spent = request.params[3].get_bool();
+    }
```
Comment (promag) [inline src/rpc/blockchain.cpp]: Disallow `include_spent` if `include_mempool`:
```cpp
    if (fMempool && include_spent) {
        throw JSONRPCError(RPC_INVALID_PARAMETER, "...");
    }
```
If you add this, update test below.

File: src/validation.cpp [hunk 4]
```diff
@@ -955,6 +948,23 @@ bool GetTransaction(const uint256 &hash, CTransactionRef &txOut, const Consensus
     return false;
 }
 
+/**
+ * Return transaction in txOut, and if it was found inside a block,
+ * its hash is placed in hashBlock. Look in the mempool first.
+*/
+bool GetTransaction(const uint256 &hash, CTransactionRef &txOut, const Consensus::Params& consensusParams, uint256 &hashBlock, bool fAllowSlow)
```
Comment (promag) [inline src/validation.cpp]: snake_case.

**Comment by jtimon** [comment]
Needs rebase and has nits but since I didn't heard any enthusiasm for more functionality regarding -txindex and @sipa was negative about it, I am closing this for now.
@promag thanks for the review and your nits will be taken into account if I ever reopen this.
